# Task 1 — Data Exploration

**Project:** Analysis of Bag Tax Data in U.S. Stores  
**Course:** MGT 8180 — Sustainable Management 
Analytics  
**Professor Name:** Dr. Nesreen El-Rayes 
**Student:** Osman Fatih Karaca  
**Dataset:** Bag Tax Data   
**Source:** https://gastate.view.usg.edu/d2l/le/content/3534044/viewContent/74711043/View 

-----

## Objectives of This Notebook

1. Load the dataset into Python  
2. Inspect its structure (rows, columns, data types)  
3. Identify missing or inconsistent data  
4. Clean the dataset for analysis  
5. Provide a summary of the cleaned data

-----

## Tools Used
- **Python (pandas, numpy)** — data loading and cleaning  
- **Jupyter Notebook in VS Code** — analysis environment 


######  Note to self — Markdown cheat sheet

- `#` = biggest heading (title)
- `##` = second-level heading
- `###` to `######` = smaller headings (smallest is `######`)
- `**bold text**` = **bold**
- `*italic text*` = *italic*
- `---` = horizontal line / divider
- `- item` = bullet point
- `> text` = blockquote

In [19]:
# Import the necessary libraries
import pandas as pd
import numpy as np

# Confirm that the libraries are imported successfully
print("pandas version:", pd.__version__)
print("numpy version:", np.__version__)

pandas version: 2.3.3
numpy version: 2.3.5


In [20]:
# Load the Bag Tax Dataset from the CSV file
df = pd.read_csv('../../Data/Bag_Tax_Data.csv', encoding='utf-8-sig')
# Display the first few rows of the dataset to confirm it loaded correctly
df.head(10) 

,File ID,Account,Date From,Date To,Bag Count,Amount Collected,Amount Due,Amount Retained,Create Date,Vendor Name,City,State,Zip code,Location1
0,95540,5151,10/1/2024,10/30/2024,35,1.75,1.40,0.35,10/30/2024,"David's Bridal, Inc.",NaN,PA,19428,"PA 19428\n(40.0797, -75.30055)"
1,95534,5080,9/26/2024,10/25/2024,615,30.75,24.60,6.15,10/26/2024,ELITE RETAILERS LLC,Silver Spring,MD,20901,"Silver Spring, MD 20901\n(39.02585, -77.01502)"
2,95533,738,9/25/2024,10/25/2024,2480,124.00,99.20,24.80,10/25/2024,"cpss, inc t/a 3 star beer and wine",Gaithersburg,MD,20879,"Gaithersburg, MD 20879\n(39.16574, -77.18153)"
3,95378,57,9/4/2024,10/1/2024,241481,12074.05,9659.24,2414.81,10/17/2024,Harris Teeter,NaN,NC,28105,"NC 28105\n(35.11822, -80.72025)"
4,95135,3763,9/2/2024,9/29/2024,5816,290.80,232.64,58.16,10/2/2024,PICCA CHICKEN INC,Germantown,MD,20874,"Germantown, MD 20874\n(39.18272, -77.27048)"
5,95134,3590,9/2/2024,9/29/2024,8256,412.80,330.24,82.56,10/2/2024,"Sardis Chicken II, Inc.",Gaithersburg,MD,20877,"Gaithersburg, MD 20877\n(39.14033, -77.19373)"
6,95160,3770,9/2/2024,9/30/2024,1550,77.50,62.00,15.50,10/5/2024,KIMO LLC,Gaithersburg,MD,20877,"Gaithersburg, MD 20877\n(39.14033, -77.19373)"
7,95157,84,9/2/2024,9/29/2024,19308,965.40,772.32,193.08,10/4/2024,"The Fresh Market, Inc.",NaN,NC,27401,"NC 27401\n(36.06908, -79.79505)"
8,95133,3918,9/2/2024,9/29/2024,5330,266.50,213.20,53.30,10/2/2024,"Sardis Chicken VI, Inc",Takoma Park,MD,20912,"Takoma Park, MD 20912\n(38.98126, -77.00988)"
9,95136,3919,9/2/2024,9/29/2024,2931,146.55,117.24,29.31,10/2/2024,"Sardis Fusion, Inc",Olney,MD,20832,"Olney, MD 20832\n(39.15052, -77.06888)"


In [21]:
# Inspect the structure of the dataset

# 1. Shape - number of rows and columns
print("Dataset shape(rows, columns):", df.shape)
print()
# 2. Columns - list of columns names
print("Column Names:")
print(list(df.columns))
print()
# 3. Data Types - Data types of each columns
print("Data Types:")
print(df.dtypes)
print()
# 4. Missing Values - Checking for missin values in each columns
print("Missing values per column:")
print(df.isna().sum())

Dataset shape(rows, columns): (71973, 14)

Column Names:
['File ID', 'Account', 'Date From', 'Date To', 'Bag Count', 'Amount Collected', 'Amount Due', 'Amount Retained', 'Create Date', 'Vendor Name', 'City', 'State', 'Zip code', 'Location1']

Data Types:
File ID               int64
Account               int64
Date From            object
Date To              object
Bag Count             int64
Amount Collected    float64
Amount Due          float64
Amount Retained     float64
Create Date          object
Vendor Name          object
City                 object
State                object
Zip code              int64
Location1            object
dtype: object

Missing values per column:
File ID                 0
Account                 0
Date From               0
Date To                 0
Bag Count               0
Amount Collected        0
Amount Due              0
Amount Retained         0
Create Date             0
Vendor Name             0
City                22956
State                   0

In [22]:
# Show full missing value counts for each columns (not truncated)
pd.set_option('display.max_rows', None) #show all rows in the output
print("Missing values per column :")
print(df.isna().sum())
pd.reset_option('display.max_rows') #reset to default (10) rows

Missing values per column :
File ID                 0
Account                 0
Date From               0
Date To                 0
Bag Count               0
Amount Collected        0
Amount Due              0
Amount Retained         0
Create Date             0
Vendor Name             0
City                22956
State                   0
Zip code                0
Location1               0
dtype: int64


### Missing data observations 

- The dataset contains **71,973 records** and **14 columns**.
- Only one column has missing values: **City** (22,956 missing values, ~32% of records). But they have locations.
- All other columns are complete (0 missing values).

### Decision

I will **keep all rows** and **not impute** the missing City values. Reason:
- City is not required for any analysis task (Bag Count, Amount Collected, vendor analysis, time trends).
- The **State** and **Zip code** columns are fully populated, providing complete geographic information.
- All transactions in this dataset occurred at retail establishments, regardless of vendor billing address.

In [23]:
# Since there are no missing values, I realized Date format is not in datetime format. I will convert it to datetime format for easier analysis later on.
# Step 1: Convert 'Date' Column to Datetime Format
df['date From'] = pd.to_datetime(df['Date From'], errors='coerce') # Convert to datetime, coerce errors to NaT
df['Date To'] = pd.to_datetime(df['Date To'], errors='coerce') 
# Step 2: What is the date range of the dataset?
print("=" * 60)
print("DATE RANGE CHECK")
print("=" * 60)
print(f"Date From — earliest: {df['Date From'].min()}")
print(f"Date From — latest:   {df['Date From'].max()}")
print(f"Date To   — earliest: {df['Date To'].min()}")
print(f"Date To   — latest:   {df['Date To'].max()}")

DATE RANGE CHECK
Date From — earliest: 1/1/2003
Date From — latest:   9/9/2019
Date To   — earliest: 1931-12-31 00:00:00
Date To   — latest:   2106-09-30 00:00:00


In [24]:
# Re-convert all date columns from text to datetime
df['Date From'] = pd.to_datetime(df['Date From'], errors='coerce')
df['Date To'] = pd.to_datetime(df['Date To'], errors='coerce')
df['Create Date'] = pd.to_datetime(df['Create Date'], errors='coerce')

# Verify the conversion worked — check the data type
print("Data types after conversion:")
print(df[['Date From', 'Date To', 'Create Date']].dtypes)
print()
print("=" * 60)
print("DATE RANGE — VERIFIED")
print("=" * 60)
print(f"Date From — earliest: {df['Date From'].min()}")
print(f"Date From — latest:   {df['Date From'].max()}")
print(f"Date To   — earliest: {df['Date To'].min()}")
print(f"Date To   — latest:   {df['Date To'].max()}")


Data types after conversion:
Date From      datetime64[ns]
Date To        datetime64[ns]
Create Date    datetime64[ns]
dtype: object

DATE RANGE — VERIFIED
Date From — earliest: 1931-03-01 00:00:00
Date From — latest:   2024-10-01 00:00:00
Date To   — earliest: 1931-12-31 00:00:00
Date To   — latest:   2106-09-30 00:00:00


In [25]:
# Show the actual bad rows
print("=" * 60)
print("BAD ROWS — Date From before 2012-01-01")
print("=" * 60)
bad_old = df[df['Date From'] < '2012-01-01']
print(f"Total bad rows: {len(bad_old)}")
print()
print(bad_old[['File ID', 'Date From', 'Date To', 'Vendor Name', 'Bag Count']].head(10))
print()
print("=" * 60)
print("BAD ROWS — Date To after 2025-12-31")
print("=" * 60)
bad_future = df[df['Date To'] > '2025-12-31']
print(f"Total bad rows: {len(bad_future)}")
print()
print(bad_future[['File ID', 'Date From', 'Date To', 'Vendor Name', 'Bag Count']].head(10))

BAD ROWS — Date From before 2012-01-01
Total bad rows: 58

       File ID  Date From    Date To                      Vendor Name  \
71915    12485 2011-12-10 2014-02-15            Aj and Aj Corporation   
71916     4641 2011-12-01 2012-10-27             Schurman Fine Papers   
71917    11954 2011-12-01 2013-12-31                       FAAN, INC.   
71918    10814 2011-10-01 2013-10-31               johnny & prum Inc.   
71919    11643 2011-10-01 2012-11-30                           11-Jul   
71920    10225 2011-09-16 2013-10-08            Aj and Aj Corporation   
71921    10249 2011-09-01 2013-09-30                   Dale Music Co.   
71922     3315 2011-07-01 2011-08-31              la brasa rotisserie   
71923    17505 2011-07-01 2014-10-31  Sugarloaf Mountain Vineyard LLC   
71924    15296 2011-06-29 2014-07-26                  HANCOCK FABRICS   

       Bag Count  
71915       1402  
71916      11015  
71917       1045  
71918       1779  
71919       6710  
71920        921  
7192

In [26]:
# ============================================================
# Cleaning Step: Remove rows with invalid dates
# ============================================================
# Justification:
# - Montgomery County's open data portal states the dataset 
#   contains records "since this law went into effect" (Jan 1, 2012)
# - Rows with Date From before 2012-01-01 are outside this scope
# - Rows with Date To far in the future (2027+) are data entry errors

# Save the original count for comparison
original_count = len(df)
print(f"Original record count: {original_count:,}")
print()

# Define what counts as a valid record
valid_mask = (
    (df['Date From'] >= '2012-01-01') &     # Tax started here
    (df['Date To'] <= '2025-12-31') &       # Reasonable upper bound
    (df['Date From'] <= df['Date To'])      # From must be before To
)

# Apply the filter — keep only valid rows
df_clean = df[valid_mask].copy()

# Report what was removed
removed = original_count - len(df_clean)
print(f"Rows removed:           {removed:,}")
print(f"Records after cleaning: {len(df_clean):,}")
print(f"Percentage retained:    {len(df_clean)/original_count*100:.2f}%")

Original record count: 71,973

Rows removed:           193
Records after cleaning: 71,780
Percentage retained:    99.73%


In [27]:
# ============================================================
# Verification: Confirm bad rows are gone
# ============================================================

print("BEFORE cleaning:")
print(f"  Date From — earliest: {df['Date From'].min()}")
print(f"  Date From — latest:   {df['Date From'].max()}")
print(f"  Date To   — earliest: {df['Date To'].min()}")
print(f"  Date To   — latest:   {df['Date To'].max()}")
print()
print("AFTER cleaning:")
print(f"  Date From — earliest: {df_clean['Date From'].min()}")
print(f"  Date From — latest:   {df_clean['Date From'].max()}")
print(f"  Date To   — earliest: {df_clean['Date To'].min()}")
print(f"  Date To   — latest:   {df_clean['Date To'].max()}")

BEFORE cleaning:
  Date From — earliest: 1931-03-01 00:00:00
  Date From — latest:   2024-10-01 00:00:00
  Date To   — earliest: 1931-12-31 00:00:00
  Date To   — latest:   2106-09-30 00:00:00

AFTER cleaning:
  Date From — earliest: 2012-01-01 00:00:00
  Date From — latest:   2024-10-01 00:00:00
  Date To   — earliest: 2012-01-03 00:00:00
  Date To   — latest:   2024-10-30 00:00:00


In [28]:
# ============================================================
# Save the cleaned dataset for use in Tasks 2, 3, 4
# ============================================================
import os

# Create the clean folder if it doesn't exist
clean_folder = '../../Data/clean'
os.makedirs(clean_folder, exist_ok=True)

# Save the cleaned data
output_path = f'{clean_folder}/Bag_Tax_Clean.csv'
df_clean.to_csv(output_path, index=False)

print(f"✅ Cleaned data saved to: {output_path}")
print(f"   Rows: {len(df_clean):,}")
print(f"   Columns: {len(df_clean.columns)}")

✅ Cleaned data saved to: ../../Data/clean/Bag_Tax_Clean.csv
   Rows: 71,780
   Columns: 15


In [29]:
print(df_clean.columns.tolist())

['File ID', 'Account', 'Date From', 'Date To', 'Bag Count', 'Amount Collected', 'Amount Due', 'Amount Retained', 'Create Date', 'Vendor Name', 'City', 'State', 'Zip code', 'Location1', 'date From']


In [30]:
# Remove the accidental duplicate 'date From' column
if 'date From' in df_clean.columns:
    df_clean = df_clean.drop(columns=['date From'])
    print("✅ Removed duplicate 'date From' column")

# Verify columns now
print(f"\nColumn count: {len(df_clean.columns)}")
print("\nAll columns:")
for col in df_clean.columns:
    print(f"  • {col}")

✅ Removed duplicate 'date From' column

Column count: 14

All columns:
  • File ID
  • Account
  • Date From
  • Date To
  • Bag Count
  • Amount Collected
  • Amount Due
  • Amount Retained
  • Create Date
  • Vendor Name
  • City
  • State
  • Zip code
  • Location1


In [31]:
# Re-save with the corrected columns
output_path = '../../Data/clean/Bag_Tax_Clean.csv'
df_clean.to_csv(output_path, index=False)
print(f"\n✅ Re-saved cleaned data: {len(df_clean):,} rows × {len(df_clean.columns)} columns")


✅ Re-saved cleaned data: 71,780 rows × 14 columns


In [32]:
# ============================================================
# TASK 1 — FINAL DATASET SUMMARY
# ============================================================

print("=" * 60)
print("MONTGOMERY COUNTY BAG TAX — CLEANED DATASET SUMMARY")
print("=" * 60)
print()

print("DATASET SIZE")
print(f"   Records (after cleaning):   {len(df_clean):,}")
print(f"   Columns:                     {len(df_clean.columns)}")
print(f"   Original records:            {len(df):,}")
print(f"   Records removed:             {len(df) - len(df_clean):,}")
print()

print("DATE RANGE")
print(f"   Earliest filing:             {df_clean['Date From'].min().date()}")
print(f"   Latest filing:               {df_clean['Date From'].max().date()}")
years_covered = df_clean['Date From'].dt.year.nunique()
print(f"   Years covered:               {years_covered}")
print()

print("VENDORS")
print(f"   Unique vendor names:         {df_clean['Vendor Name'].nunique():,}")
print(f"   Unique vendor accounts:      {df_clean['Account'].nunique():,}")
print()

print("BAG ACTIVITY")
print(f"   Total bags reported:         {df_clean['Bag Count'].sum():,}")
print(f"   Total tax collected:         ${df_clean['Amount Collected'].sum():,.2f}")
print(f"   Remitted to county (80%):    ${df_clean['Amount Due'].sum():,.2f}")
print(f"   Retained by vendors (20%):   ${df_clean['Amount Retained'].sum():,.2f}")
print()

print("GEOGRAPHIC COVERAGE")
print(f"   Unique states (vendor HQ):   {df_clean['State'].nunique()}")
print(f"   Most common state:           {df_clean['State'].value_counts().index[0]} ({df_clean['State'].value_counts().iloc[0]:,} records)")
print()

print("KEY VARIABLES")
key_vars = ['File ID', 'Account', 'Date From', 'Date To', 'Bag Count',
            'Amount Collected', 'Amount Due', 'Amount Retained',
            'Vendor Name', 'State', 'Zip code']
for v in key_vars:
    print(f"   • {v}")

MONTGOMERY COUNTY BAG TAX — CLEANED DATASET SUMMARY

DATASET SIZE
   Records (after cleaning):   71,780
   Columns:                     14
   Original records:            71,973
   Records removed:             193

DATE RANGE
   Earliest filing:             2012-01-01
   Latest filing:               2024-10-01
   Years covered:               13

VENDORS
   Unique vendor names:         1,738
   Unique vendor accounts:      1,776

BAG ACTIVITY
   Total bags reported:         843,798,620
   Total tax collected:         $42,189,851.87
   Remitted to county (80%):    $33,751,948.24
   Retained by vendors (20%):   $8,437,903.63

GEOGRAPHIC COVERAGE
   Unique states (vendor HQ):   37
   Most common state:           MD (50,515 records)

KEY VARIABLES
   • File ID
   • Account
   • Date From
   • Date To
   • Bag Count
   • Amount Collected
   • Amount Due
   • Amount Retained
   • Vendor Name
   • State
   • Zip code


---

##  Task 1 - Summary of Work Completed

### Steps performed
1. **Loaded** `Bag_Tax_Data.csv` from the Data folder using pandas.
2. **Inspected** the dataset — 71,973 records, 14 columns.
3. **Identified missing values** — only the `City` column had missing data (22,956 rows, ~32%). 
   Decision: kept all rows, since `State` and `Zip code` provide complete geographic information.
4. **Identified inconsistent dates:**
   - 58 rows with `Date From` before January 1, 2012 (the law's effective date)
   - 6 rows with `Date To` in clearly impossible future years (2048, 2106, etc.)
   - Justification: Montgomery County's open data portal explicitly states the dataset 
     contains records *"since this law went into effect"* on January 1, 2012.
5. **Cleaned** the dataset by removing 193 invalid rows (0.27% of data).
6. **Verified** the cleaning — minimum date is now January 1, 2012; maximum date is October 30, 2024.
7. **Saved** the cleaned dataset to `Data/clean/Bag_Tax_Clean.csv` for use in Tasks 2–5.

### Final dataset summary
- **Records:** 71,780 (after cleaning)
- **Date range:** January 1, 2012 – October 1, 2024 (~12.75 years)
- **Unique vendors:** 1,738
- **Total bags reported:** 843,798,620
- **Total tax collected:** $42,189,851.87
   - Remitted to county (80%): $33,751,948.24
   - Retained by vendors (20%): $8,437,903.63

### Source citation
Montgomery County Open Data Portal — Bag Tax dataset.  
Bill 8-11 (Carryout Bag Tax Law), effective January 1, 2012.  
Source: data.montgomerycountymd.gov/Finance-Tax-Property/Bag-Tax/xnjh-vgc4
